File: CGStudio/PGCore/Math/Primitive3d/Circle3d.cs
namespace PG.Core.Math
{
    public class Circle3d
    {
        private Vector3d center;
        private Vector3d uvec;
        private Vector3d vvec;

        public Vector3d Center
        {
            get { return center; }
            set { this.center = value; }
        }

        public Circle3d() :
            this(new Vector3d(0,0,0), new Vector3d(1,0,0), new Vector3d(0,1,0))
        {
        }

        public Circle3d(Vector3d center, Vector3d uvec, Vector3d vvec)
        {
            this.center = center;
            this.uvec = uvec;
            this.vvec = vvec;
        }

        public Vector3d GetPosition(double u)
        {
            var angle = u * 2.0 * System.Math.PI;
            var x = uvec * System.Math.Cos(angle);
            var y = vvec * System.Math.Sin(angle);
            return center + x + y;
        }

        public Vector3d Normal()
        {
            return uvec.Normalized.Cross(vvec.Normalized);
        }
    }
}

File: CGLib/Crystal/Math/Circle3d.cpp
#include "Circle3d.h"

#include "Tolerance.h"

#include <cmath>

using namespace Crystal::Math;

template<typename T>
Circle3d<T>::Circle3d() :
	center(Vector3dd(0,0,0)),
	uvec(Vector3dd(1,0,0)),
	vvec(Vector3dd(0,1,0))
{}

template<typename T>
Circle3d<T>::Circle3d(const Vector3d<T>& center, const Vector3d<T>& uvec, const Vector3d<T>& vvec) :
	center(center),
	uvec(uvec),
	vvec(vvec)
{}

template<typename T>
Vector3d<T> Circle3d<T>::getPosition(const T u) const
{
	const auto uu = u * T(2.0) * T(PI);
	return center + ::cos(uu) * uvec + ::sin(uu) * vvec;
}

template<typename T>
Vector3d<T> Circle3d<T>::getNormal() const
{
	return glm::normalize( glm::cross(uvec, vvec) );
}


template class Circle3d<float>;
template class Circle3d<double>;